shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,790 @@
-  // Determines whether the vertex is allowed to open the grain by searching depth first
-  // for a path of active role assignments leading from the grain owner to the vertex.
+function normalize(membranedPermissionSetMap, desiredPermissions) {
+  // `membranedPermissionSetMap` represents the permissions flowing from a grain owner to a vertex,
+  // exactly in the format of the output of `backpropagateVertex()`.
+  //
+  // `desiredPermissions` is the set of permissions that we are currently interested in.
+  //
+  // Returns a set of clauses in a minimal disjunctive normal form, representing the possible
+  // ways that the desired permissions could be achieved. The result is a map from hashed clause
+  // to objects of the form { clause: Clause, tokensUsed: [String] }.
+
+  const result = {};
 
-  var edges = collectEdges(db, vertex);
-  if (edges.grainDoesNotExist || edges.disallowedAnonymousAccess) {
-    return false;
+  function fullCover(chosen) {
+    if (Object.keys(chosen).length == 0) return false;
+
+    const accum = new PermissionSet([]);
+    for (const hashedClause in chosen) {
+      accum.add(membranedPermissionSetMap[hashedClause].permissions);
+    }
+
+    return desiredPermissions.isSubsetOf(accum);
   }
-  if (edges.grainIsPublic) {
-    return true;
+
+  const chooseAll = {};
+  for (const hashedClause in membranedPermissionSetMap) {
+    const pclause = membranedPermissionSetMap[hashedClause];
+    chooseAll[hashedClause] = { clause: pclause.membrane, tokensUsed: pclause.tokensUsed };
   }
-  var edgesByRecipient = edges.edgesByRecipient;
-  var owner = "OwningAccount";
-  if (!edges.terminalEdge) { return false; }
-  if (owner == edges.terminalEdge.sharer) {
-    return true;
+
+  if (!fullCover(chooseAll)) {
+    return {};
   }
 
-  var openerIdentityId = edges.terminalEdge.sharer;
-  var stackedUsers = {};
-  stackedUsers[openerIdentityId] = true;
-  var userStack = [openerIdentityId];
-
-  while (userStack.length > 0) {
-    var recipient = userStack.pop();
-    var edges = edgesByRecipient[recipient];
-    if (edges) {
-      for (var ii = 0; ii < edges.length; ++ii) {
-        var inEdge = edges[ii];
-        var sharer = inEdge.sharer;
-        if (sharer == owner) {
-          return true;
+  const stack = [chooseAll];
+  // Array of Map<hashedClause, { clause: Clause, tokensUsed: [String] }>.
+
+  while (stack.length > 0) {
+    const chosen = stack.pop();
+    let minimal = true;
+    for (const removedHashedClause in chosen) {
+      const newChosen = {};
+      for (const hashedClause in chosen) {
+        if (hashedClause !== removedHashedClause) {
+          newChosen[hashedClause] = chosen[hashedClause];
         }
-        if (!stackedUsers[sharer]) {
-          userStack.push(sharer);
-          stackedUsers[sharer] = true;
+      }
+
+      if (fullCover(newChosen)) {
+        minimal = false;
+        // TODO(perf): keep track of values we've already explored.
```

Comment:
Note that the immediately straightforward way to compute `normalize()` takes can take p^n time, where p is number of permissions and n is the number of ConditionalPermissionSets in the input.

I've updated the function so that it groups together ConditionalPermissionSets that correspond to the same PermissionSet. This way, when we do the "minimal cover" computation, all of the subsets we deal with are unique. I think this makes the computation (modulo the TODO(perf) about remembering states we've visited already) come to O(2^m) where m is the number of unique PermissionSets in the input permissions flow.

Because it can still blow up, I've added an iteration counter that bails after a while. I suspect that one would need to willingly construct pathological examples in order to get this to happen.